### DISCUSSION TITLE
Scoped preflight for gradual migration

### DISCUSSION THREAD

**Discussion by frolic** [tw-in-js/twind#141] [Q&A] [answered]
I'm adding twind to an existing React project with its own CSS (via emotion). I noticed that twind styles are well isolated to the components that use them, but once the component unmounts, some of the base styles (resets, I think) linger on the page.

Is there a recommended approach to cleanly removing these when the component is unmounted?

**Comment by frolic** [comment]
To clarify, I think I want to keep the preflight in place. I tried `setup({ preflight: false })` and that had more side effects than I'd like. I don't mind if I have to fix up any side effects of using Twind with other things on the page, just want to ensure that Twind styles clean up as the components unmount.

**Comment by sastan** [comment]
Why do you want them to be removed? They are in a style element. Your emotion styles are kept as well.

**Comment by frolic** [comment]
There's is some overlap between the twind base styles/resets and the existing styles on the page that I am trying to resolve. We're introducing twind incrementally, and the overlap in styles makes it hard to do so.

Maybe what I'm looking for is a virtual sheet that is mounted alongside each component that uses twind?

**Comment by frolic** [comment]
An example of a twind base style that gets injected when I'm using it with a component:
![image](https://user-images.githubusercontent.com/508855/110036247-92fdc800-7cf1-11eb-8072-a4ea2f7afa8e.png)

When that component is mounted on the page, anything of those elements with a margin get reset. So, for example, a set of plain `<p>` tags styled by Bootstrap will all lose their bottom margin.

I am okay with this problem/overlap existing when the component is mounted, but it remains when the component is unmounted.

**Comment by sastan** [comment]
I see! Got you. The preflight causes issues. You want a scoped preflight. That may be difficult because the additional scope like `.tw h1` increases the [specificity](https://specificity.keegan.st/) and therefore overrules the utility classes.  [See this example](https://esm.codes/#aW1wb3J0IHsgdHcsIHNldHVwIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQnCmltcG9ydCB7IGRvbVNoZWV0IH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQvc2hlZXRzJwoKc2V0dXAoewogIC8vIHNjb3BlIGFsbCBwcmVmbGlnaHQgc3R5bGVzIHRvIGEgY29udGFpbmVyIHdpdGggJ3R3JyBhcyBjbGFzcwogIHByZWZsaWdodDogKHsgaHRtbCwgYm9keSwgLi4uY3NzIH0pID0+ICh7CiAgICAnLnR3JzogewogICAgICAnJic6IHsgLi4uaHRtbCwgLi4uYm9keSwgbWFyZ2luOiAnMCcgfSwKICAgICAgLi4uY3NzLAogICAgfQogIH0pLAogIC8vIEp1c3QgZm9yIHRoaXMgZGVtbwogIHNoZWV0OiBkb21TaGVldCgpCn0pCgpkb2N1bWVudC5ib2R5LmlubmVySFRNTCA9IGAKICA8aDE+Tm8gVHdpbmQgcHJlZmxpZ2h0PC9oMT4KICA8bWFpbiBjbGFzcz0idHcgJHt0d2BoLXNjcmVlbiBiZy1wdXJwbGUtNDAwIGZsZXggaXRlbXMtY2VudGVyIGp1c3RpZnktY2VudGVyYH0iPgogICAgPGgxIGNsYXNzPSIke3R3YGZvbnQtYm9sZCB0ZXh0KGNlbnRlciA1eGwgd2hpdGUgc206Z3JheS04MDAgbWQ6cGluay03MDApYH0iPlRoaXMgaXMgVHdpbmQhPC9oMT4KICA8L21haW4+CmA=)

To solve that you must then increase the specificity of all Twind generated classes for example using the `override` variant. [Here is an idea](https://esm.codes/#aW1wb3J0IHsgdHcgYXMgX3R3LCBzZXR1cCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kJwppbXBvcnQgeyBkb21TaGVldCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kL3NoZWV0cycKCnNldHVwKHsKICAvLyBzY29wZSBhbGwgcHJlZmxpZ2h0IHN0eWxlcyB0byBhIGNvbnRhaW5lciB3aXRoICd0dycgYXMgY2xhc3MKICBwcmVmbGlnaHQ6ICh7IGh0bWwsIGJvZHksIC4uLmNzcyB9KSA9PiAoewogICAgJy50dyc6IHsKICAgICAgJyYnOiB7IC4uLmh0bWwsIC4uLmJvZHksIG1hcmdpbjogJzAnIH0sCiAgICAgIC4uLmNzcywKICAgIH0KICB9KSwKICAvLyBKdXN0IGZvciB0aGlzIGRlbW8KICBzaGVldDogZG9tU2hlZXQoKQp9KQoKLy8gRW5zdXJlIGFsbCB1dGlsaXRpZXMgaGF2ZSBhIGhpZ2hlciBzcGVjaWZpY2l0eSB0aGVuIHRoZSBwcmVmbGlnaHQgc3R5bGVzIApjb25zdCB0dyA9ICh0b2tlbikgPT4gJ3R3ICcgKyBfdHdgb3ZlcnJpZGU6KCR7dG9rZW59KWAKCmRvY3VtZW50LmJvZHkuaW5uZXJIVE1MID0gYAogIDxoMT5ObyBUd2luZCBwcmVmbGlnaHQ8L2gxPgogIDxtYWluIGNsYXNzPSIke3R3YGgtc2NyZWVuIGJnLXB1cnBsZS00MDAgZmxleCBpdGVtcy1jZW50ZXIganVzdGlmeS1jZW50ZXJgfSI+CiAgICA8aDEgY2xhc3M9IiR7dHdgZm9udC1ib2xkIHRleHQoY2VudGVyIDV4bCB3aGl0ZSBzbTpncmF5LTgwMCBtZDpwaW5rLTcwMClgfSI+VGhpcyBpcyBUd2luZCE8L2gxPgogIDwvbWFpbj4KYA==)

Sadly Twind does not have a configurable way to increase the specificity for all generated classes.

**Comment by frolic** [comment]
Oh perfect, this might be exactly what I need. Will report back. Thank you for the quick help!

**Comment by frolic** [comment]
Oof, yeah, the specificity does not play nicely with other things. For example, all of react-select's default styles get reset when in a `.tw` container.

Any other ideas?

**Comment by sastan** [comment]
No preflight – but a base style that is applied to all elements that are styled with Twind: [Example](https://esm.codes/#aW1wb3J0IHsgdHcgYXMgX3R3LCBzZXR1cCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kJwppbXBvcnQgeyBkb21TaGVldCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kL3NoZWV0cycKCnNldHVwKHsKICAvLyBEZWZpbmUgYmFzZSBzdHlsZSBmb3IgYWxsIFR3aW5kIHJ1bGVzCiAgcHJlZmxpZ2h0OiAoeyBodG1sLCBib2R5LCAnKiw6OmJlZm9yZSw6OmFmdGVyJzogYm9yZGVyQm94LCB9KSA9PiAoewogICAgJy50dyc6IHsKICAgICAgLy8gT3JkZXIgaXMgaW1wb3J0YW50CiAgICAgIC4uLmJvZHksIC4uLmh0bWwsIC4uLmJvcmRlckJveCwKICAgICAgbWFyZ2luOiAnMCcsIHBhZGRpbmc6ICcwJywKICAgICAgJyY6OmJlZm9yZSwmOjphZnRlcic6IGJvcmRlckJveAogICAgfQogIH0pLAogIC8vIEp1c3QgZm9yIHRoaXMgZGVtbwogIHNoZWV0OiBkb21TaGVldCgpCn0pCgovLyBVc2UgYmFzZSBzdHlsZSBmb3IgYWxsIFR3aW5kIHJ1bGVzCmNvbnN0IHR3ID0gKC4uLnRva2VucykgPT4gJ3R3ICcgKyBfdHcoLi4udG9rZW5zKQoKZG9jdW1lbnQuYm9keS5pbm5lckhUTUwgPSBgCiAgPGgxPk5vIFR3aW5kIHByZWZsaWdodDwvaDE+CiAgPG1haW4gY2xhc3M9IiR7dHdgaC1zY3JlZW4gYmctcHVycGxlLTQwMCBmbGV4IGl0ZW1zLWNlbnRlciBqdXN0aWZ5LWNlbnRlcmB9Ij4KICAgIDxoMSBjbGFzcz0iJHt0d2Bmb250LWJvbGQgdGV4dChjZW50ZXIgNXhsIHdoaXRlIHNtOmdyYXktODAwIG1kOnBpbmstNzAwKWB9Ij5UaGlzIGlzIFR3aW5kITwvaDE+CiAgPC9tYWluPgpg)

Or with hashed classes there is no need for a custom `tw` function: [Example](https://esm.codes/#aW1wb3J0IHsgdHcsIHNldHVwIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQnCmltcG9ydCB7IGRvbVNoZWV0IH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQvc2hlZXRzJwoKc2V0dXAoewogIGhhc2g6IHRydWUsCiAgLy8gRGVmaW5lIGJhc2Ugc3R5bGUgZm9yIGFsbCBUd2luZCBydWxlcwogIHByZWZsaWdodDogKHsgaHRtbCwgYm9keSwgJyosOjpiZWZvcmUsOjphZnRlcic6IGJvcmRlckJveCwgfSkgPT4gKHsKICAgICdbY2xhc3MqPXR3LV0nOiB7CiAgICAgIC8vIE9yZGVyIGlzIGltcG9ydGFudAogICAgICAuLi5ib2R5LCAuLi5odG1sLCAuLi5ib3JkZXJCb3gsCiAgICAgIG1hcmdpbjogJzAnLCBwYWRkaW5nOiAnMCcsCiAgICAgICcmOjpiZWZvcmUsJjo6YWZ0ZXInOiBib3JkZXJCb3gKICAgIH0KICB9KSwKICAvLyBKdXN0IGZvciB0aGlzIGRlbW8KICBzaGVldDogZG9tU2hlZXQoKQp9KQoKZG9jdW1lbnQuYm9keS5pbm5lckhUTUwgPSBgCiAgPGgxPk5vIFR3aW5kIHByZWZsaWdodDwvaDE+CiAgPG1haW4gY2xhc3M9IiR7dHdgaC1zY3JlZW4gYmctcHVycGxlLTQwMCBmbGV4IGl0ZW1zLWNlbnRlciBqdXN0aWZ5LWNlbnRlcmB9Ij4KICAgIDxoMSBjbGFzcz0iJHt0d2Bmb250LWJvbGQgdGV4dChjZW50ZXIgNXhsIHdoaXRlIHNtOmdyYXktODAwIG1kOnBpbmstNzAwKWB9Ij5UaGlzIGlzIFR3aW5kITwvaDE+CiAgPC9tYWluPgpg)

**Reply by hastebrot** [reply]
These examples helped me to use Twind scoped in ObservableHQ notebooks without resetting the styles provided by ObservableHQ.

Demp: https://next.observablehq.com/@hastebrot/using-twind@181

BTW: Thanks for this amazing library!

**Comment by sastan** [comment]
@holic Did the last comment help?

**Comment by frolic** [comment]
Thanks for checking back in. I think it helped somewhat. There’s this disappearing bottom border thing I discovered when there’s no preflight that I’m trying to debug, but needed to move on to something else yesterday. Will be poking at this some more on Monday.

**Comment by frolic** [comment]
[Here's](https://esm.codes/#aW1wb3J0IHsgdHcsIHNldHVwIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQnCgpzZXR1cCh7CiAgcHJlZmxpZ2h0OiBmYWxzZSwKfSkKCmRvY3VtZW50LmJvZHkuaW5uZXJIVE1MID0gYAogIDxoMT5ObyBUd2luZCBwcmVmbGlnaHQ8L2gxPgogIDxtYWluIGNsYXNzPSIke3R3YGgtc2NyZWVuIGJnLXB1cnBsZS00MDAgZmxleCBpdGVtcy1jZW50ZXIganVzdGlmeS1jZW50ZXJgfSI+CiAgICA8aDEgY2xhc3M9IiR7dHdgZm9udC1ib2xkIHRleHQoY2VudGVyIDV4bCB3aGl0ZSBzbTpncmF5LTgwMCBtZDpwaW5rLTcwMCkgYm9yZGVyLWItMiBib3JkZXItYmx1ZS01MDBgfSI+VGhpcyBpcyBUd2luZCE8L2gxPgogIDwvbWFpbj4KYA==) the issue I was seeing. With preflight disabled, some of the borders don't appear. If you remove `preflight: false`, you'll see the bottom border on the h1. Any idea what's going on here?

As far as I can tell from the Chrome inspector, these borders _should_ appear - they aren't grayed out/struckthrough like I would expect if they were being overridden by other styles.

![image](https://user-images.githubusercontent.com/508855/110263089-76b89000-7f6a-11eb-8c2f-3ab7bbf5e766.png)

**Comment by frolic** [comment]
Ah, maybe it's the lack of this in the base style?
![image](https://user-images.githubusercontent.com/508855/110263204-d2831900-7f6a-11eb-8ee7-780b1e68c031.png)
Specifically the missing `solid` without the preflight causes the border not to appear.

If I [add `border-solid` without the preflight](https://esm.codes/#aW1wb3J0IHsgdHcsIHNldHVwIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQnCgpzZXR1cCh7CiAgcHJlZmxpZ2h0OiBmYWxzZSwKfSkKCmRvY3VtZW50LmJvZHkuaW5uZXJIVE1MID0gYAogIDxoMT5ObyBUd2luZCBwcmVmbGlnaHQ8L2gxPgogIDxtYWluIGNsYXNzPSIke3R3YGgtc2NyZWVuIGJnLXB1cnBsZS00MDAgZmxleCBpdGVtcy1jZW50ZXIganVzdGlmeS1jZW50ZXJgfSI+CiAgICA8aDEgY2xhc3M9IiR7dHdgZm9udC1ib2xkIHRleHQoY2VudGVyIDV4bCB3aGl0ZSBzbTpncmF5LTgwMCBtZDpwaW5rLTcwMCkgIGJvcmRlci1zb2xpZCBib3JkZXItYi0yIGJvcmRlci1ibHVlLTUwMGB9Ij5UaGlzIGlzIFR3aW5kITwvaDE+CiAgPC9tYWluPgpg), it shows up but seems to override the border widths so the full border appears.

It's not until I [add both `border-solid` _and_ `border-0`](https://esm.codes/#aW1wb3J0IHsgdHcsIHNldHVwIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQnCgpzZXR1cCh7CiAgcHJlZmxpZ2h0OiBmYWxzZSwKfSkKCmRvY3VtZW50LmJvZHkuaW5uZXJIVE1MID0gYAogIDxoMT5ObyBUd2luZCBwcmVmbGlnaHQ8L2gxPgogIDxtYWluIGNsYXNzPSIke3R3YGgtc2NyZWVuIGJnLXB1cnBsZS00MDAgZmxleCBpdGVtcy1jZW50ZXIganVzdGlmeS1jZW50ZXJgfSI+CiAgICA8aDEgY2xhc3M9IiR7dHdgZm9udC1ib2xkIHRleHQoY2VudGVyIDV4bCB3aGl0ZSBzbTpncmF5LTgwMCBtZDpwaW5rLTcwMCkgIGJvcmRlci1zb2xpZCBib3JkZXItMCBib3JkZXItYi0yIGJvcmRlci1ibHVlLTUwMGB9Ij5UaGlzIGlzIFR3aW5kITwvaDE+CiAgPC9tYWluPgpg) that it works as expected.

**Comment by sastan** [comment]
Have you checked this comment: https://github.com/tw-in-js/twind/issues/134#issuecomment-791036921

It is working there. [Example](https://esm.codes/#aW1wb3J0IHsgdHcgYXMgX3R3LCBzZXR1cCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kJwppbXBvcnQgeyBkb21TaGVldCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kL3NoZWV0cycKCnNldHVwKHsKICAvLyBEZWZpbmUgYmFzZSBzdHlsZSBmb3IgYWxsIFR3aW5kIHJ1bGVzCiAgcHJlZmxpZ2h0OiAoeyBodG1sLCBib2R5LCAnKiw6OmJlZm9yZSw6OmFmdGVyJzogYm9yZGVyQm94LCB9KSA9PiAoewogICAgJy50dyc6IHsKICAgICAgLy8gT3JkZXIgaXMgaW1wb3J0YW50CiAgICAgIC4uLmJvZHksIC4uLmh0bWwsIC4uLmJvcmRlckJveCwKICAgICAgbWFyZ2luOiAnMCcsIHBhZGRpbmc6ICcwJywKICAgICAgJyY6OmJlZm9yZSwmOjphZnRlcic6IGJvcmRlckJveAogICAgfQogIH0pLAogIC8vIEp1c3QgZm9yIHRoaXMgZGVtbwogIHNoZWV0OiBkb21TaGVldCgpCn0pCgovLyBVc2UgYmFzZSBzdHlsZSBmb3IgYWxsIFR3aW5kIHJ1bGVzCmNvbnN0IHR3ID0gKC4uLnRva2VucykgPT4gJ3R3ICcgKyBfdHcoLi4udG9rZW5zKQoKZG9jdW1lbnQuYm9keS5pbm5lckhUTUwgPSBgCiAgPGgxPk5vIFR3aW5kIHByZWZsaWdodDwvaDE+CiAgPG1haW4gY2xhc3M9IiR7dHdgaC1zY3JlZW4gYmctcHVycGxlLTQwMCBmbGV4IGl0ZW1zLWNlbnRlciBqdXN0aWZ5LWNlbnRlcmB9Ij4KICAgIDxoMSBjbGFzcz0iJHt0d2Bmb250LWJvbGQgdGV4dChjZW50ZXIgNXhsIHdoaXRlIHNtOmdyYXktODAwIG1kOnBpbmstNzAwKSBib3JkZXItYi0yIGJvcmRlci1ibHVlLTUwMGB9Ij5UaGlzIGlzIFR3aW5kITwvaDE+CiAgPC9tYWluPgpg)

**Comment by sastan** [comment]
BTW: I changed the title because we may be on to something here. That could evolve into an own module like twind/scoped

```js
import { tw, setup } from 'twind/scoped'
```

**Comment by frolic** [comment]
> Have you checked this comment: [#134 (comment)](https://github.com/tw-in-js/twind/issues/134#issuecomment-791036921)
> 
> It is working there. [Example](https://esm.codes/#aW1wb3J0IHsgdHcgYXMgX3R3LCBzZXR1cCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kJwppbXBvcnQgeyBkb21TaGVldCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kL3NoZWV0cycKCnNldHVwKHsKICAvLyBEZWZpbmUgYmFzZSBzdHlsZSBmb3IgYWxsIFR3aW5kIHJ1bGVzCiAgcHJlZmxpZ2h0OiAoeyBodG1sLCBib2R5LCAnKiw6OmJlZm9yZSw6OmFmdGVyJzogYm9yZGVyQm94LCB9KSA9PiAoewogICAgJy50dyc6IHsKICAgICAgLy8gT3JkZXIgaXMgaW1wb3J0YW50CiAgICAgIC4uLmJvZHksIC4uLmh0bWwsIC4uLmJvcmRlckJveCwKICAgICAgbWFyZ2luOiAnMCcsIHBhZGRpbmc6ICcwJywKICAgICAgJyY6OmJlZm9yZSwmOjphZnRlcic6IGJvcmRlckJveAogICAgfQogIH0pLAogIC8vIEp1c3QgZm9yIHRoaXMgZGVtbwogIHNoZWV0OiBkb21TaGVldCgpCn0pCgovLyBVc2UgYmFzZSBzdHlsZSBmb3IgYWxsIFR3aW5kIHJ1bGVzCmNvbnN0IHR3ID0gKC4uLnRva2VucykgPT4gJ3R3ICcgKyBfdHcoLi4udG9rZW5zKQoKZG9jdW1lbnQuYm9keS5pbm5lckhUTUwgPSBgCiAgPGgxPk5vIFR3aW5kIHByZWZsaWdodDwvaDE+CiAgPG1haW4gY2xhc3M9IiR7dHdgaC1zY3JlZW4gYmctcHVycGxlLTQwMCBmbGV4IGl0ZW1zLWNlbnRlciBqdXN0aWZ5LWNlbnRlcmB9Ij4KICAgIDxoMSBjbGFzcz0iJHt0d2Bmb250LWJvbGQgdGV4dChjZW50ZXIgNXhsIHdoaXRlIHNtOmdyYXktODAwIG1kOnBpbmstNzAwKSBib3JkZXItYi0yIGJvcmRlci1ibHVlLTUwMGB9Ij5UaGlzIGlzIFR3aW5kITwvaDE+CiAgPC9tYWluPgpg)

I did but Typescript was complaining about those spreads. 

![image](https://user-images.githubusercontent.com/508855/110361130-72d15000-7ff4-11eb-9ca9-b6c6908009e7.png)

Even if I add `// @ts-ignore` here, I'm seeing ~the same results.

[Here's](https://esm.codes/#aW1wb3J0IHsgdHcgYXMgX3R3LCBzZXR1cCwgY3NzIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQvY3NzJwppbXBvcnQgeyBkb21TaGVldCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kL3NoZWV0cycKCnNldHVwKHsKICAvLyBEZWZpbmUgYmFzZSBzdHlsZSBmb3IgYWxsIFR3aW5kIHJ1bGVzCiAgcHJlZmxpZ2h0OiAoeyBodG1sLCBib2R5LCAnKiw6OmJlZm9yZSw6OmFmdGVyJzogYm9yZGVyQm94LCB9KSA9PiAoewogICAgJy50dyc6IHsKICAgICAgLy8gT3JkZXIgaXMgaW1wb3J0YW50CiAgICAgIC4uLmJvZHksIC4uLmh0bWwsIC4uLmJvcmRlckJveCwKICAgICAgbWFyZ2luOiAnMCcsIHBhZGRpbmc6ICcwJywKICAgICAgJyY6OmJlZm9yZSwmOjphZnRlcic6IGJvcmRlckJveAogICAgfQogIH0pLAogIC8vIEp1c3QgZm9yIHRoaXMgZGVtbwogIHNoZWV0OiBkb21TaGVldCgpCn0pCgovLyBVc2UgYmFzZSBzdHlsZSBmb3IgYWxsIFR3aW5kIHJ1bGVzCmNvbnN0IHR3ID0gKC4uLnRva2VucykgPT4gJ3R3ICcgKyBfdHcoLi4udG9rZW5zKQoKY29uc3Qgc3R5bGVzID0gY3NzYAogIEBhcHBseSBoLXNjcmVlbiBiZy1wdXJwbGUtNDAwIGZsZXggaXRlbXMtY2VudGVyIGp1c3RpZnktY2VudGVyOwogIAogID4gaDEgewogICAgQGFwcGx5IGZvbnQtYm9sZCB0ZXh0KGNlbnRlciA1eGwgd2hpdGUgc206Z3JheS04MDAgbWQ6cGluay03MDApOwogICAgQGFwcGx5IGJvcmRlci1iLTIgYm9yZGVyLWJsdWUtNTAwOwogIH0KYDsKCmRvY3VtZW50LmJvZHkuaW5uZXJIVE1MID0gYAogIDxoMT5ObyBUd2luZCBwcmVmbGlnaHQ8L2gxPgogIDxtYWluIGNsYXNzPSIke3R3KHN0eWxlcyl9Ij4KICAgIDxoMT5UaGlzIGlzIFR3aW5kITwvaDE+CiAgPC9tYWluPgpg) a better reproduction of how I'm using twind and where it's not working.

Maybe it's because the `borderBox` key was losing its `*` selector? [This](https://esm.codes/#aW1wb3J0IHsgdHcgYXMgX3R3LCBzZXR1cCwgY3NzIH0gZnJvbSAnaHR0cHM6Ly9jZG4uc2t5cGFjay5kZXYvdHdpbmQvY3NzJwppbXBvcnQgeyBkb21TaGVldCB9IGZyb20gJ2h0dHBzOi8vY2RuLnNreXBhY2suZGV2L3R3aW5kL3NoZWV0cycKCnNldHVwKHsKICAvLyBEZWZpbmUgYmFzZSBzdHlsZSBmb3IgYWxsIFR3aW5kIHJ1bGVzCiAgcHJlZmxpZ2h0OiAoeyBodG1sLCBib2R5LCAnKiw6OmJlZm9yZSw6OmFmdGVyJzogYm9yZGVyQm94LCB9KSA9PiAoewogICAgJy50dyc6IHsKICAgICAgLy8gT3JkZXIgaXMgaW1wb3J0YW50CiAgICAgIC4uLmJvZHksIC4uLmh0bWwsIC4uLmJvcmRlckJveCwKICAgICAgbWFyZ2luOiAnMCcsIHBhZGRpbmc6ICcwJywKICAgICAgJyosJjo6YmVmb3JlLCY6OmFmdGVyJzogYm9yZGVyQm94CiAgICB9CiAgfSksCiAgLy8gSnVzdCBmb3IgdGhpcyBkZW1vCiAgc2hlZXQ6IGRvbVNoZWV0KCkKfSkKCi8vIFVzZSBiYXNlIHN0eWxlIGZvciBhbGwgVHdpbmQgcnVsZXMKY29uc3QgdHcgPSAoLi4udG9rZW5zKSA9PiAndHcgJyArIF90dyguLi50b2tlbnMpCgpjb25zdCBzdHlsZXMgPSBjc3NgCiAgQGFwcGx5IGgtc2NyZWVuIGJnLXB1cnBsZS00MDAgZmxleCBpdGVtcy1jZW50ZXIganVzdGlmeS1jZW50ZXI7CiAgCiAgPiBoMSB7CiAgICBAYXBwbHkgZm9udC1ib2xkIHRleHQoY2VudGVyIDV4bCB3aGl0ZSBzbTpncmF5LTgwMCBtZDpwaW5rLTcwMCk7CiAgICBAYXBwbHkgYm9yZGVyLWItMiBib3JkZXItYmx1ZS01MDA7CiAgfQpgOwoKZG9jdW1lbnQuYm9keS5pbm5lckhUTUwgPSBgCiAgPGgxPk5vIFR3aW5kIHByZWZsaWdodDwvaDE+CiAgPG1haW4gY2xhc3M9IiR7dHcoc3R5bGVzKX0iPgogICAgPGgxPlRoaXMgaXMgVHdpbmQhPC9oMT4KICA8L21haW4+CmA=) seems to do the trick.

---

Edit: Nope, that's bad too, because then all downstream components get unset (e.g. a Bootstrap input field loses its border). Maybe we need all descendant selectors/classes to be prefixed with `.tw` too?

I might just live with the simple `preflight: false` for now and work around the border issue with `border-solid border-0`. I can't think of a great way to isolate the styles.

**Comment by sastan** [comment]
> I did but Typescript was complaining about those spreads.

You need to cast them:

```ts
import type { CSSRules } from 'twind'

setup({
  // Define base style for all Twind rules
  preflight: ({ html, body, '*,::before,::after': borderBox, }) => ({
    '.tw': {
      // Order is important
      ...(body as CSSRules),
      ...(html  as CSSRules),
      ...(borderBox  as CSSRules),
      margin: '0',
      padding: '0',
      '&::before,&::after': borderBox
    }
  }),
})
```

> Maybe we need all descendant selectors/classes to be prefixed with .tw too?

Then we ran into specificity issues again. `.tw button` > `.border-1`

Maybe we should get the `apply` route you already started. Mixing different CSS frameworks is really tough.

I will stop looking for solutions. But if you need any help or have questions please reach out to me.

I would like to move this issue into discussions because I do not really see this as an issue with twind or something that could be solved by twind. 

Would that be OK with you?

**Comment by frolic** [comment, marked as answer]
That seems reasonable! Thanks for all the help 🙏

**Comment by sastan** [comment]
Just a reminder for later: we could use the [:where](https://developer.mozilla.org/en-US/docs/Web/CSS/:where) selector which always has 0 specificity.

**Comment by aweber1** [comment]
@holic @sastan 
FWIW, I needed scoped preflight as well but took it in a slightly different direction. I have a `twind`-styled micro-app that is "embedded" within a site with existing styling, so want to scope the twind preflight to just my embedded components.  The code below will prefix _all_ the preflight CSS rules with a custom class name that is scoped to my embedded app/components.

Works great for my use case, hope others might find value!

```
// styles.ts
import { tw as baseTw, setup, CSSRules, CSSRuleValue } from 'twind';

const resetWrapperClassName = 'my-reset-wrapper';

setup({
  preflight: prefixPreflight,
});

export const tw = (...tokens: any) => `${resetWrapperClassName} ${baseTw(...tokens)}`;

function prefixPreflight({ html, body, '*,::before,::after': borderBox, ...otherCssRules }: CSSRules) {
  const prefixedRules: Record<string, CSSRuleValue> = {};
  Object.keys(otherCssRules).forEach((cssRuleKey) => {
    // some rules contain multiple selectors, e.g. "h1,h2,h3,h4,h5,h6"
    // we need to split those into separate prefixed rules and assign the same css properties to each.
    cssRuleKey.split(',').forEach((cssSelector) => {
      const prefixedSelector = `.${resetWrapperClassName} ${cssSelector}`;
      // merge rules if some already exist
      if (prefixedRules[prefixedSelector]) {
        prefixedRules[prefixedSelector] = {
          ...(prefixedRules[prefixedSelector] as any),
          ...(otherCssRules[cssRuleKey] as any),
        };
      } else {
        prefixedRules[prefixedSelector] = otherCssRules[cssRuleKey];
      }
    });
  });

  return {
    [`.${resetWrapperClassName}`]: {
      // Order is important
      ...(body as CSSRules),
      ...(html as CSSRules),
      ...(borderBox as CSSRules),
      margin: '0',
      padding: '0',
      '&::before,&::after': borderBox,
    },
    [`.${resetWrapperClassName} *, .${resetWrapperClassName} *::before, .${resetWrapperClassName} *::after`]: borderBox,
    ...prefixedRules,
  };
}

```

**Reply by danielweck** [reply]
Reminds me of this:
https://github.com/thomaspark/scoper/blob/5de04619ca20221663e480faa7b7288bcef24e2e/scoper.js#L10-L34

**Reply by marbemac** [reply]
Doesn't this suffer from the specificity issues though?

`.sl` is our prefix in this case:

<img width="296" alt="Screen Shot 2021-06-20 at 3 19 31 PM" src="https://user-images.githubusercontent.com/847542/122687253-0eb94d00-d1db-11eb-89b4-67679b64265a.png">

Because the preflight rules are prefixed, they end up with a higher specificity than the tailwind classes:

<img width="338" alt="Screen Shot 2021-06-20 at 3 21 52 PM" src="https://user-images.githubusercontent.com/847542/122687284-40321880-d1db-11eb-98b2-5d62a8ce495b.png">

In this case our button ends up with `0` padding from the preflight, rather than the `px-3` we intend.

**Reply by marbemac** [reply]
What could be pretty snazzy if if we could "scope/prefix" everything - e.g. if `scope` is `.my-app`, all preflight rules AND all tailwind classes that are added to the document are prefixed by that scope, thus solving the specificity issue 🤔.

**Comment by lveillard** [comment]
Hello! We would love to have a TwindWrapper that enables twind only for its children and grand children. 

The examples in esm.codes have disppeared and a long time has passed so I was wondering if there is a better /simple way to create this wrapper that isolates the twind runtime classes.

Thanks!

**Reply by lveillard** [reply]
This works but i have absolutely no idea what it does. I just used temporally the Frankenstein family name, mixed everything I've found, cleaned it a bit and added types 😅

```ts

const sheetWithRootSelector = (sheet: Sheet, selector: string): Sheet => ({
  ...sheet,
  insert: (rule: string, index: number, ruleObj: SheetRule) => {
    const mediaMatches = rule.match(/^@[^{]+\{/);
    const mediaQuery = mediaMatches ? mediaMatches[0] : '';
    const cleanedRule = rule
      .replace(mediaQuery, '')
      .replace(/\b(,\s*)?(html|body)\b/g, '')
      .replace(/^\s*,/, '');

    let prefixedRule = `${selector} ${cleanedRule}`;

    if (/\b(html|body)\b/.test(rule)) {
      prefixedRule = `${selector}, ${prefixedRule}`;
    }

    const finalRule = mediaQuery ? `${mediaQuery}${prefixedRule}` : prefixedRule;

    return sheet.insert(finalRule, index, ruleObj);
  },
});

const TwindWrapper = ({ children }: { children: React.ReactNode }) => {
  const [isLoaded, setIsLoaded] = useState(false);

  useEffect(() => {
    const config = defineConfig({
      presets: [presetAutoprefix(), presetTailwind()],
    });

    setup(config, sheetWithRootSelector(cssom(), '#twindWrapper'));
    setIsLoaded(true);
  }, []);

  if (!isLoaded) {
    return null;
  }
  return <div id="twindWrapper">{children}</div>;
};
```
